File: src/csv/checkins.js
const csv = require('@fast-csv/format')

module.exports = async function outputCheckins (out, hugo) {
  const stream = csv.format({
    headers: [
      'Date',
      'Name',
      'Country',
      'Region',
      'Locality',
      'Address',
      'Latitude',
      'Longitude',
      'Tags'
    ]
  })
  stream.pipe(out)

  for await (const { meta } of hugo.getAll({ keepOriginal: true })) {
    if (!meta.properties || !meta.properties.checkin) {
      continue
    }

    stream.write([
      meta.date.getTime(),
      meta.properties.checkin.properties.name,
      meta.properties.checkin.properties['country-name'],
      meta.properties.checkin.properties.region,
      meta.properties.checkin.properties.locality,
      meta.properties.checkin.properties['street-address'],
      meta.properties.checkin.properties.latitude,
      meta.properties.checkin.properties.longitude,
      meta.tags
    ])
  }

  stream.end()
}

File: eagle/checkins.go
package eagle

import (
	"sort"
	"time"
)

type Checkin struct {
	Date      time.Time `csv:"date"`
	Latitude  float64   `csv:"latitude"`
	Longitude float64   `csv:"longitude"`
	Name      string    `csv:"name"`
	Locality  string    `csv:"locality"`
	Region    string    `csv:"region"`
	Country   string    `csv:"country"`
}

func (c *Checkin) Multiformats() map[string]interface{} {
	props := map[string]interface{}{
		"latitude":  c.Latitude,
		"longitude": c.Longitude,
	}

	if c.Name != "" {
		props["name"] = c.Name
	}

	if c.Locality != "" {
		props["locality"] = c.Locality
	}

	if c.Region != "" {
		props["region"] = c.Region
	}

	if c.Country != "" {
		props["country-name"] = c.Country
	}

	return map[string]interface{}{
		"type":       "h-adr",
		"properties": props,
	}
}

type Checkins []*Checkin

func (cc Checkins) Sort() Checkins {
	sort.Slice(cc, func(i, j int) bool {
		return cc[i].Date.Before(cc[j].Date)
	})
	return cc
}
